tutorials/ServerConfig.md
Various configuration examples to run huntJS applications
==================================

This folders contains config files for any 3rd party programs,
that are used to run nodejs applications build on Hunt.js

Main idea behind this is to allow Hunt.js application to be ran on behalf
of limited user account. This is more save, than running hunt.js applications
as root user, but it prevents us from listening on ports with numbers greater
than 1000 - because on Linux only root user can run applications that listens
on standard ports of 80 and 443.

We assume you have Linux machine.

[firewall.sh](https://github.com/vodolaz095/hunt/blob/master/examples/serverConfigsExamples/firewall.sh)
==================================

Shell script to properly setup the firewall - open 22, 80, 443 port

[pound.cfg](https://github.com/vodolaz095/hunt/blob/master/examples/serverConfigsExamples/pound.cfg)
==================================

Configuration file for [Pound reverse proxy and load balancer](http://www.apsis.ch/pound).
So, we can run the Pound service as root on 80 port as http server, and on 443 port
as https server, with http backend on port 3000, that is a hunt application, ran
on behalf of user with limited account.
The strong sides of this approach is very high security and performance of
Pound, and the bad side is issues with websockets connectivity in socket.io module.

[nginx.conf](https://github.com/vodolaz095/hunt/blob/master/examples/serverConfigsExamples/nginx.conf)
==================================

Configuration file for [nginx web server and reverse proxy](http://nginx.org/).
So, we can run the Nginx service as root on 80 port as http server, and on 443 port
as https server, with http backend on port 3000, that is a hunt application, ran
on behalf of user with limited account. Also we serve static assets from `public/`
directory of hunt.js application by the means of nginx.
Also the websockets are supported

[huntjs@.service](https://github.com/vodolaz095/hunt/blob/master/examples/serverConfigsExamples/huntjs%40.service)
==================================

[Systemd linux initialization system](https://en.wikipedia.org/wiki/Systemd) allows us
to start HuntJs application as limited user on system startup.
After editing this file and placing it to `/etc/systemd/system`, we can control our huntjs 
application like this:

```shell

  $ su -c 'systemctl restart huntjs@myUserName.service'
  $ su -c 'systemctl start huntjs@myUserName.service'
  $ su -c 'systemctl status huntjs@myUserName.service'
  $ su -c 'systemctl enable huntjs@myUserName.service'
  $ su -c 'systemctl disable huntjs@myUserName.service'

```

[huntjs](https://github.com/vodolaz095/hunt/blob/master/examples/serverConfigsExamples/huntjs)
==================================
Config example to run HuntJS application by means of `chkconfig` - 
[http://linuxcommand.org/man_pages/chkconfig8.html](http://linuxcommand.org/man_pages/chkconfig8.html)

CHANGELOG.md
# v0.4.12
Improvements in documentation and tutorials.

# v0.4.11
Fix issue with `Hunt.extendTelnet` not returning `Hunt` instance
Improvements in code of example.
Updated dependencies to more recent ones. 
Fix rare race conditions in `Hunt.preload` middleware.

# v0.4.9
[https://www.npmjs.com/package/validator](Validator) is exported as `hunt.validator`.

# v0.4.8
Updated dependencies to more recent ones

# v0.4.7
We can configure socket.io by via config object with this parameters all of this parameters
[https://github.com/Automattic/engine.io#methods-1](https://github.com/Automattic/engine.io#methods-1)

# v0.4.6
Updated dependencies to more recent ones

# v0.4.5
Fix rare issue with malformed user profiles in session storage.
User#email setter converts the email to lower case in all places where it is used. (Sorry for Cyrillic domains)
Break socket.io connection in Mozilla browser like described here 
[https://bugzilla.mozilla.org/show_bug.cgi?id=712329](ttps://bugzilla.mozilla.org/show_bug.cgi?id=712329)
Fix rare bug in local strategy for reseting password with malformed users account

# v0.4.4
User#notifyByEmail now have the user profile in the context of email message template.
Fix rare issue with reseting password with missing `keychain.welcomeLink`

# v0.4.3
Fix issues for sending emails with undefined callback

# v0.4.2
Fix for clearing temporary files resulted from processing uploaded files via `Hunt.multipart` middleware 

# v0.4.1
Feature - we can use hunt.emit('broadcast', message) to emit socket.io messages with different payload,
defined in `message.type`. The backward compatibility with 3.x.x branch exists.
Minor updates in dependencies

# v0.4.0
We use the Mongoose 4.x branch now. The CHANGELOG.md is created.


lib/misc/config.js
'use strict';

/**
 * @class config
 * @classdesc
 * Configuration object, that is passed to Hunt constructor
 *
 * @property {string} hostUrl - url to host of current application,
 * used in oAuth authentication strategies and for building redirects.
 * For example - http://example.org/
 *
 * @property {string} env - application environment, if not set, populated
 * from enviromental value of NODE_ENV. Default is `development`, the values
 * of `staging` and `production` can be used
 *
 * @property {number} port - port for webserver process(es) to bind to.
 * Note: on *nix machines ports with number bellow 1000 are only bindable
 * by applicaiton ran as root user.
 *
 * @property {(string|null)} address - what address to bind to.
 * Default is '0.0.0.0' - all IPv4 addresses.
 * The address is populated from environment address of HUNTJS_ADDR.
 * HuntJS supports both IPv4 and IPv6 addresses as described
 * {@link http://code.danyork.com/2011/01/21/how-to-use-node-js-with-ipv6/ here}.
 *
 * @property {boolean} enableMongoose - enable mongo support via mongoose orm, default is true - enabled
 * @property {string} mongoUrl - connection string for mongoose ORM,
 * default is 'mongodb://localhost/hunt_dev'.
 * It can be automatically populated from enviromental values of `MONGO_URL`,
 * `MONGOSOUP_URL`,`MONGOHQ_URL`,`MONGOLAB_URL`.
 * Connection string has this {@link http://docs.mongodb.org/manual/reference/connection-string/ syntax }
 *
 * @property {boolean} enableMongooseUsers - enable mongodb based user object, build by pattern of Active Record.
 * Default is true - enabled
 *
 * @property {string} redisUrl - connection string for redis, default is
 * 'redis://localhost:6379', can be populated from enviromental variables
 * of 'REDISTOGO_URL', 'OPENREDIS_URL', 'REDISCLOUD_URL',
 *  'REDISGREEN_URL', 'REDIS_URL'
 *
 * @property {string} sequelizeUrl - URI for relational database connection. Is used in {@link https://www.npmjs.org/package/hunt-sequilize | Hunt-sequilize} plugin
 *
 * @property {string} public - directory for assets - css, images,
 * client side javascripts, that are served by {@link http://www.senchalabs.org/connect/static.html | connect static middleware}
 * for example, 'public': __dirname+'/public'
 *
 * @property {string} views - directory for {@link http://expressjs.com/api.html#app-settings | templates},
 * used by template engine for rendering static HTML pages on server side.
 * See this {@link https://huntjs.herokuapp.com/documentation/tutorial-templating.html | tutorial} for details
 *
 * @property {string} emailConfig - configuration string for {@link https://npmjs.org/package/nodemailer/ | email delivery system },
 * used for notify users.
 * When left blank, the direct transport is used - it is quite slow, and the emails
 * are usually marked as spam.
 * Also you can set up SMTP options directly like mentioned here {@link https://github.com/andris9/nodemailer#setting-up-smtp }
 * Examples:
 * 'nodemailer://somebody@gmail.com:somepassword@gmail',
 * 'nodemailer://postmaster@teksi.ru:7tu8z3zsaa421@Mailgun',
 *
 * @property {Boolean} dialog - enable dialog REST-api for private messages
 * @property {Boolean} huntKey - enable authorization by huntKey as GET query, or POST/PUT/DELETE body parameter. Default is false - disabled
 * @property {Boolean} huntKeyHeader - enable authorization by huntKey as custom header, default is false - disabled
 *
 * @property {Boolean} disableCsrf - disable csrf protection, default is false - CSRF protection is enabled. See {@link https://www.npmjs.org/package/csurf } for details. Also this works from the box with AngularJS $resourse and $http
 *
 * @property {String} adminEmail - email address of server administrator. If present, every error will be reported to this email with plenty of additional information and error stack.
 *
 * @property {number} maxWorkers - limit the number of webserver/background processes cluster spawns. Default - 1 per CPU core.
 *
 * @property {passport} passport - {@link http://passportjs.org | passport.js}
 * configuration object, used for setting the methods users authorize,
 *
 * @property {boolean} uploadFiles -  allow upload of files by HTTP-POST,
 * {@link http://expressjs.com/api.html#req.files},
 * default is false, disabled, and for a good cause {@link https://groups.google.com/forum/#!topic/nodejs/6KOlfk5cpcM}
 *
 * @property {object} io - configuration options for {@link http://socket.io/}. Default is `{'enabled':false}`, make `{'enabled':true} to enable. We can use this parameters for setup - {@link https://github.com/Automattic/engine.io#methods-1}. If we pass `true`, the sane default values will be set.
 * @property {string} favicon - path to favicon
 * @property {string} secret - string to preseed session hashes and do other security related things
 * @property {object} telnetServer - configuration options for RAI telnet server {@link https://github.com/andris9/rai#starting-a-server}
 */

require('colors');
var os = require('os'),
  crypto = require('crypto'),
  url = require('url');

function md5(str) {
  return crypto.createHash('md5').update(str).digest('hex').toString();
}

module.exports = exports = function (configToUse) {
  var config = configToUse || {};
  config.env = config.env || process.env.NODE_ENV || 'development';
  config.port = config.port || parseInt(process.env.PORT) || 3000;
  config.secret = config.secret || process.env.secret || null;
  config.adminEmail = config.adminEmail || process.env.ADMIN_EMAIL || null;
  config.address = config.address || process.env.HUNTJS_ADDR || '0.0.0.0';

  if (!config.adminEmail) {
    console.error(('ConfigManager: config.adminEmail is missing! Set the config parameter of adminEmail or environment value of ADMIN_EMAIL to be able to recieve email reports with errors!').yellow)
  }
  if (!config.secret) {
    console.error(('ConfigManager: config.secret is missing! Generating the secret on the fly...').yellow);
    config.secret = (md5(JSON.stringify(os)));
  }

  if (config.hostUrl) {
    console.log(('ConfigManager: hostUrl is set to ' + config.hostUrl).green);
  }

  if (!config.hostUrl) {
    if (process.env.hostUrl) {
      config.hostUrl = process.env.hostUrl;
    } else {
      var hostname = os.hostname();
      if (config.env === 'development') {
        hostname = 'localhost';
      }
      config.hostUrl = 'http://' + hostname + ':' + config.port + '/';
    }
    console.log(('ConfigManager: assuming hostUrl is ' + config.hostUrl).yellow);
  }

//trying to setup mail dispatcher
  if (!config.emailConfig) {
    config.emailConfig = process.env.emailConfig;
  }

  if (!config.favicon) {
    config.favicon = __dirname + '/favicon.ico';
  }

//trying to get redis connection parameters
  if (config.redisUrl) {
    console.log('ConfigManager: REDIS'.green + ' extracted connection parameters from config object!');
  } else {
    [
      'REDISTOGO_URL',
      'OPENREDIS_URL',
      'REDISCLOUD_URL',
      'REDISGREEN_URL',
      'REDIS_URL'
    ].map(function (redisKey) {
        if (process.env[redisKey]) {
          console.log('ConfigManager: REDIS'.green + ' extracted connection parameters from process.env.' + redisKey + ' !');
          config.redisUrl = process.env[redisKey];
        }
      });
    if (!config.redisUrl) {
      config.redisUrl = 'redis://localhost:6379';
      console.log('ConfigManager: REDIS'.yellow + ' used default redis connection parameters - redis://localhost:6379');
    }
  }

//autoloading mongoose ORM and mongoose users model
  if (config.enableMongoose === undefined) {
    config.enableMongoose = true;
  }
  if (config.enableMongoose && config.enableMongooseUsers === undefined) {
    config.enableMongooseUsers = true;
  }

  if (config.mongoUrl) {
    console.log('ConfigManager: MONGO'.green + ' extracted connection parameters from config object!');
  }

  if (config.enableMongoose && !config.mongoUrl) {
    [
      'MONGOLAB_URI',
      'MONGOSOUP_URL',
      'MONGOHQ_URL',
      'MONGO_URL'
    ].map(function (mongoKey) {
        if (process.env[mongoKey]) {
          console.log('ConfigManager: MONGO'.green + ' extracted connection parameters from process.env.' +
            mongoKey + ' !');
          config.mongoUrl = process.env[mongoKey];
        }
      });

    if (!config.mongoUrl) {
      console.log('ConfigManager: MONGO'.yellow + ' used default connection settings  - mongodb://localhost/hunt_dev');
      config.mongoUrl = 'mongodb://localhost/hunt_dev';
    }
  }
//sequilize setup

  if (config.sequelizeUrl) {
    console.log('ConfigManager: Sequilize'.green + ' extracted connection parameters from config object');
  }

  [
    'DATABASE_URL', //heroku postgress
    'CLEARDB_DATABASE_URL', //cleardb mysql
    'SQL_URL' //something user provided
  ].map(function (sqlProviderName) {
      if (process.env[sqlProviderName] && !config.sequelizeUrl) {
        config.sequelizeUrl = process.env[sqlProviderName];
        console.log('ConfigManager: Sequilize'.green + ' extracted connection parameters from process.env.' + sqlProviderName + ' !');
      }
    });

//setting socket.io parameters
//with same sane values
// see for details - https://github.com/Automattic/engine.io#methods-1
  if(config.io === true) {
    config.io = {'enabled': true};
  }
  config.io = config.io || {};
  config.io.pingTimeout = config.io.pingTimeout || 5000;
  config.io.pingInterval = config.io.pingInterval || 10000;
  config.io.allowUpgrades = config.io.allowUpgrades !== undefined  ? config.io.allowUpgrades : true;
  config.io.transports = config.io.transports || ['polling', 'websocket'];


//setting authorization strategies parameters
  config.passport = config.passport || {};
  config.passport.sessionExpireAfterSeconds = config.passport.sessionExpireAfterSeconds || 180;
  config.passport.local = (config.passport.local === undefined) ? true : config.passport.local;

  if (config.passport.local) {

    config.passport.signUpByEmail = config.passport.signUpByEmail || false;
    if (config.passport.signUpByEmail) {
      config.passport.verifyEmail = config.passport.verifyEmail || false;
      if (config.passport.verifyEmail && !config.passport.verifyEmailTemplate) {
        throw new Error('Error setting template for email used for verifying user\'s account!');
      }
    }
    config.passport.resetPassword = config.passport.resetPassword || false;
    if (config.passport.resetPassword && !config.passport.resetPasswordEmailTemplate) {
      throw new Error('Error setting template for email used for reseting password for user\'s account!');
    }
  }

//disable by default OpenID strategies
//we do not use population from environment, because they have
//simple ON/OFF settings
  config.passport.yahoo = config.passport.yahoo || false;
  config.passport.steam = config.passport.steam || false;
  config.passport.paypal = config.passport.paypal || false;
  config.passport.intuit = config.passport.intuit || false;


//getting oauth strategies settings from application environment
  [
    'GOOGLE_CLIENT_ID', 'GOOGLE_CLIENT_SECRET',
    'GITHUB_CLIENT_ID', 'GITHUB_CLIENT_SECRET',
    'TWITTER_CONSUMER_KEY', 'TWITTER_CONSUMER_SECRET',
    'FACEBOOK_APP_ID', 'FACEBOOK_APP_SECRET',
    'LINKEDIN_API_KEY', 'LINKEDIN_SECRET_KEY',
    'VK_APP_ID', 'VK_APP_SECRET'
  ].map(function (a) {
      config.passport[a] = config.passport[a] || process.env[a] || null;
    });


//scaling
  var numCPUs = os.cpus().length;
  config.maxWorkers = Math.min(numCPUs, config.maxWorkers || 16);

//setting default template engine
  config.templateEngine = config.templateEngine || 'hogan';

//making config for telnet server
  config.telnetServer = config.telnetServer || {};


//sanity checks
  var parsedRedis = url.parse(config.redisUrl);
  if (!parsedRedis || parsedRedis.protocol !== 'redis:') {
    throw new Error('Wrong redisUrl in config!');
  }

  if (config.mongoUrl) {
    var parsedMongo = url.parse(config.mongoUrl);
    if (!parsedMongo || parsedMongo.protocol !== 'mongodb:') {
      throw new Error('Wrong mongoUrl in config!');
    }
  }

  var parsedHostUrl = url.parse(config.hostUrl);
  if (!parsedHostUrl || !(parsedHostUrl.protocol === 'http:' || parsedHostUrl.protocol === 'https:')) {
    throw new Error('Wrong hostUrl in config!');
  }

  if (config.secret.length < 9) {
    throw new Error('Config.secret is to short!');
  }

  return config;
};
